src/shared/conf-parser.c

```diff
@@ -497,29 +503,54 @@ static int config_parse_many_files(
                         return -ENOMEM;
         }
 
-        /* Get inodes for all drop-ins. Later we'll verify if main config is a symlink to one of them. If so,
-         * we skip reading main config file directly. */
         STRV_FOREACH(fn, files) {
-                _cleanup_free_ struct stat *st_dropin = NULL;
+                _cleanup_fclose_ FILE *f = NULL;
+                int fd;
 
-                st_dropin = new(struct stat, 1);
-                if (!st_dropin)
-                        return -ENOMEM;
+                f = fopen(*fn, "re");
+                if (!f) {
+                        if (errno == ENOENT)
+                                continue;
 
-                r = RET_NERRNO(stat(*fn, st_dropin));
-                if (r < 0 && r != -ENOENT)
-                        return r;
+                        return -errno;
+                }
 
-                r = set_ensure_consume(&inodes, &inode_hash_ops, TAKE_PTR(st_dropin));
-                if (r < 0)
+                fd = fileno(f);
+
+                r = ordered_hashmap_ensure_put(&dropins, &config_file_hash_ops_fclose, *fn, f);
+                if (r < 0) {
+                        assert(r != -EEXIST);
                         return r;
+                }
+                assert(r > 0);
+                TAKE_PTR(f);
+
+                if (fd >= 0) {
```

Comment:
Not familiar, but in what situation this becomes negative?

Reply:
I just followed the practice here: https://github.com/systemd/systemd/blob/49549a458c761f5c75f529243b14ff5fa4950ac6/src/shared/conf-parser.c#L313-L314

But probably this can be dropped since we're certainly opening these streams through actual file?